### PR TITLE
ceph-daemon: Fix handling for symlinks on python2

### PR DESCRIPTION
**Pull request by mgfritch**
`shutil.move` correctly handles symlinks under python3, but fails when run with python2.

Signed-off-by: Michael Fritch mfritch@suse.com

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by mgfritch** [comment]
This affects the successful completion of the standalone OSD `adopt` tests (#31772).

```
+ sudo /tmp/tmp.VpugJte16r/ceph-daemon -v --image ceph/daemon-base:latest-master-devel adopt --style legacy --legacy-dir ./tmp.Cv9XoCzGGb/tmp.0fir7PWXUB --name osd.3
DEBUG:ceph-daemon:['/usr/bin/podman', 'run', '--rm', '--net=host', '-e', 'CONTAINER_IMAGE=ceph/daemon-base:latest-master-devel', '-e', 'NODE_NAME=foobar', '--entrypoint', '/usr/bin/grep', 'ceph/daemon-base:latest-master-devel', 'ceph', '/etc/passwd']
DEBUG:ceph-daemon:Running command: /usr/bin/podman run --rm --net=host -e CONTAINER_IMAGE=ceph/daemon-base:latest-master-devel -e NODE_NAME=foobar --entrypoint /usr/bin/grep ceph/daemon-base:latest-master-devel ceph /etc/passwd
DEBUG:ceph-daemon:/usr/bin/grep:stdout ceph:x:167:167:Ceph daemons:/var/lib/ceph:/sbin/nologin
DEBUG:ceph-daemon:Running command: systemctl is-enabled ceph-osd@3
DEBUG:ceph-daemon:systemctl:stderr Failed to get unit file state for ceph-osd@3.service: No such file or directory
DEBUG:ceph-daemon:Running command: systemctl is-active ceph-osd@3
DEBUG:ceph-daemon:systemctl:stdout inactive
INFO:ceph-daemon:Moving data...
DEBUG:ceph-daemon:['/usr/bin/podman', 'run', '--rm', '--net=host', '-e', 'CONTAINER_IMAGE=ceph/daemon-base:latest-master-devel', '-e', 'NODE_NAME=foobar', '--entrypoint', '/usr/bin/grep', 'ceph/daemon-base:latest-master-devel', 'ceph', '/etc/passwd']
DEBUG:ceph-daemon:Running command: /usr/bin/podman run --rm --net=host -e CONTAINER_IMAGE=ceph/daemon-base:latest-master-devel -e NODE_NAME=foobar --entrypoint /usr/bin/grep ceph/daemon-base:latest-master-devel ceph /etc/passwd
DEBUG:ceph-daemon:/usr/bin/grep:stdout ceph:x:167:167:Ceph daemons:/var/lib/ceph:/sbin/nologin
DEBUG:ceph-daemon:Move '/root/src/github.com/mgfritch/ceph/qa/standalone/test_ceph_daemon/tmp.Cv9XoCzGGb/tmp.0fir7PWXUB/var/lib/ceph/osd/ceph-3/keyring' -> '/var/lib/ceph/00000000-0000-0000-0000-ffffdeadbeef/osd.3'
DEBUG:ceph-daemon:Move '/root/src/github.com/mgfritch/ceph/qa/standalone/test_ceph_daemon/tmp.Cv9XoCzGGb/tmp.0fir7PWXUB/var/lib/ceph/osd/ceph-3/block.dev' -> '/var/lib/ceph/00000000-0000-0000-0000-ffffdeadbeef/osd.3'
DEBUG:ceph-daemon:Move '/root/src/github.com/mgfritch/ceph/qa/standalone/test_ceph_daemon/tmp.Cv9XoCzGGb/tmp.0fir7PWXUB/var/lib/ceph/osd/ceph-3/magic' -> '/var/lib/ceph/00000000-0000-0000-0000-ffffdeadbeef/osd.3'
DEBUG:ceph-daemon:Move '/root/src/github.com/mgfritch/ceph/qa/standalone/test_ceph_daemon/tmp.Cv9XoCzGGb/tmp.0fir7PWXUB/var/lib/ceph/osd/ceph-3/block' -> '/var/lib/ceph/00000000-0000-0000-0000-ffffdeadbeef/osd.3'
Traceback (most recent call last):
  File "/tmp/tmp.VpugJte16r/ceph-daemon", line 1955, in <module>
    r = args.func()
  File "/tmp/tmp.VpugJte16r/ceph-daemon", line 1515, in command_adopt
    shutil.move(data_file, data_dir_dst)
  File "/usr/lib64/python2.7/shutil.py", line 325, in move
    copy2(src, real_dst)
  File "/usr/lib64/python2.7/shutil.py", line 153, in copy2
    copyfile(src, dst)
  File "/usr/lib64/python2.7/shutil.py", line 96, in copyfile
    with open(src, 'rb') as fsrc:
IOError: [Errno 2] No such file or directory: '/root/src/github.com/mgfritch/ceph/qa/standalone/test_ceph_daemon/tmp.Cv9XoCzGGb/tmp.0fir7PWXUB/var/lib/ceph/osd/ceph-3/block'

```

File: src/ceph-daemon/ceph-daemon [hunk 1]
```diff
@@ -280,6 +280,25 @@ def make_log_dir(fsid, uid=None, gid=None):
...
+        dst = os.path.join(dst, os.path.basename(src))
+    logger.debug('Move \'%s\' -> \'%s\'' % (src, dst))
+    try:
+        return shutil.move(src, dst)
+    except IOError:
+        if os.path.islink(src):
+            # shutil.move() in python2 does not handle symlinks correctly
+            logger.debug('symlink \'%s\' -> \'%s\'' % (os.readlink(src), dst))
+            os.symlink(os.readlink(src), dst)
+            os.unlink(src)
```
Comment (sebastian-philipp) [inline src/ceph-daemon/ceph-daemon]: If `unlink` raises, `symlink` already created the destination. Is there a way to make this a tiny bit more atomic without blowing up the complexity?

**Comment by mgfritch** [comment]
> LGTM, except, aren't we trying to get rid of python2?

yeah, but the upstream tests still require both python2 and python3 ...

https://github.com/ceph/ceph/blob/61ba2d7b6688e25ebb72988516786182ef383e87/qa/standalone/test_ceph_daemon.sh#L39

File: src/ceph-daemon/ceph-daemon [hunk 2]
```diff
@@ -61,7 +61,7 @@ import uuid
 
 from distutils.spawn import find_executable
 from functools import wraps
-from glob import glob
+from glob import iglob
```
[the remark this answers is not in the document]
Comment (mgfritch) [reply src/ceph-daemon/ceph-daemon]: Avoids storing the entire list result from `glob` in memory during for loop iteration, which matters when the glob'd source dir has a very long list of files.

File: src/ceph-daemon/ceph-daemon (hunk 1 above)
Comment (mgfritch) [reply src/ceph-daemon/ceph-daemon]: I'm not sure of a better way, but I'm open to suggestions :smiley:

**Comment by sebastian-philipp** [comment]
Looking forward to the day we can remove py2.

**Comment by mgfritch** [comment]
Please re-review. The logic for handling symlinks on python2 was incorrect.

I've also added an additional commit to chown the ceph uid/gid, which finally allows for a successful adoption of an OSD using ceph-daemon!

File: src/ceph-daemon/ceph-daemon [hunk 3]
```diff
@@ -280,6 +280,47 @@ def make_log_dir(fsid, uid=None, gid=None):
     makedirs(log_dir, uid, gid, LOG_DIR_MODE)
     return log_dir
 
+def copy_file(src, dst, uid=None, gid=None):
+    """
+    Copy a file from src to dst
+    """
+    # type: (str, int, int) -> str
```
Comment (liewegas) [inline src/ceph-daemon/ceph-daemon]:
```
ceph-daemon:287: error: misplaced type annotation
```

Comment (mgfritch) [reply src/ceph-daemon/ceph-daemon]: oops .. there were two instances of this .. 

The type hint has been placed in the correct location, as well as adding the missing `str` param 

```suggestion
    # type: (str, str, int, int) -> str
```

**Pull request merged**